sdk/tracing/azotel/otel_test.go

```diff
@@ -0,0 +1,259 @@
+	assert.True(t, ok)
+	emptySpan.AddEvent("noop_event")
+
+	ctx, endSpan := azruntime.StartSpan(context.Background(), "test_span", client.Tracer(), nil)
+
+	req, err := azruntime.NewRequest(ctx, http.MethodGet, "https://www.microsoft.com/")
+	require.NoError(t, err)
+
+	startedSpan, ok := client.Tracer().SpanFromContext(req.Raw().Context())
+	assert.True(t, ok)
+	startedSpan.AddEvent("post_event")
+
+	_, err = client.Pipeline().Do(req)
+	require.NoError(t, err)
+
+	endSpan(nil)
+
+	// shut down the tracing provider to flush all spans
+	otelTP.Shutdown(context.Background())
+
+	require.Len(t, exporter.spans, 2)
+}
+
+func TestConvertSpan(t *testing.T) {
+	ts := testSpan{t: t}
+	span := convertSpan(&ts)
+
+	span.AddError(io.EOF)
+	assert.Equal(t, io.EOF, ts.recordedError)
+
+	const eventName = "event"
+	eventAttr := tracing.Attribute{
+		Key:   "key",
+		Value: "value",
+	}
+	span.AddEvent(eventName, eventAttr)
+	assert.EqualValues(t, eventName, ts.eventName)
+	require.Len(t, ts.eventOptions, 1)
+
+	span.End()
+	assert.True(t, ts.endCalled)
+
+	attr := tracing.Attribute{
+		Key:   "key",
+		Value: "value",
+	}
+	span.SetAttributes(attr)
+	require.Len(t, ts.attributes, 1)
+
+	const statusDesc = "everything is ok"
+	span.SetStatus(tracing.SpanStatusOK, statusDesc)
+	assert.EqualValues(t, tracing.SpanStatusOK, ts.statusCode)
+	assert.EqualValues(t, statusDesc, ts.statusDesc)
+}
+
+func TestConvertAttributes(t *testing.T) {
+	tests := []struct {
+		label    string
+		key      string
+		value    any
+		validate func(t *testing.T, kv attribute.KeyValue)
+	}{
+		{
+			label: "int",
+			key:   "int_key",
+			value: 123,
+			validate: func(t *testing.T, kv attribute.KeyValue) {
+				require.EqualValues(t, "INT64", kv.Value.Type().String())
+				assert.EqualValues(t, 123, kv.Value.AsInt64())
+			},
+		},
+		{
+			label: "int64",
+			key:   "int64_key",
+			value: int64(123),
+			validate: func(t *testing.T, kv attribute.KeyValue) {
+				require.EqualValues(t, "INT64", kv.Value.Type().String())
+				assert.EqualValues(t, 123, kv.Value.AsInt64())
+			},
+		},
+		{
+			label: "float64",
+			key:   "float64_key",
+			value: 3.14159,
+			validate: func(t *testing.T, kv attribute.KeyValue) {
+				require.EqualValues(t, "FLOAT64", kv.Value.Type().String())
+				assert.EqualValues(t, 3.14159, kv.Value.AsFloat64())
+			},
+		},
+		{
+			label: "bool",
+			key:   "bool_key",
+			value: true,
+			validate: func(t *testing.T, kv attribute.KeyValue) {
+				require.EqualValues(t, "BOOL", kv.Value.Type().String())
+				assert.True(t, kv.Value.AsBool())
+			},
+		},
+		{
+			label: "string",
+			key:   "string_key",
+			value: "hello",
+			validate: func(t *testing.T, kv attribute.KeyValue) {
+				require.EqualValues(t, "STRING", kv.Value.Type().String())
+				assert.EqualValues(t, "hello", kv.Value.AsString())
+			},
+		},
+		{
+			label: "float32",
+			key:   "float32_key",
+			value: float32(3.14159),
```

Comment:
I added a comment for this where you do this conversion, but it seems like we can provide a straight promotion for these to the "larger" version of the type that we do support. Although I know sometimes float promotion can be a double-edged sword.

Reply:
Agreed on rough edges WRT floats.  I'm going to leave this as-is for now which will require callers to perform the cast themselves if they don't want the `fmt.Sprintf()` fallback.  We can revisit later based on feedback.